relay-quotas/src/cache.rs

```diff
@@ -0,0 +1,409 @@
+use std::num::NonZeroUsize;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::Duration;
+
+use relay_common::time::UnixTimestamp;
+
+/// A percentage is converted to a divisor to do integer arithmetic.
+///
+/// This is done with the configured precision here.
+const PERCENT_PRECISION: usize = 10;
+
+/// A quota to be checked with the [`OpportunisticQuotaCache`].
+#[derive(Debug, Clone, Copy)]
+pub struct Quota<T> {
+    /// The quota limit.
+    pub limit: i64,
+    /// A unique identifier for the quota bucket.
+    pub key: T,
+    /// The expiry of the bucket.
+    pub expiry: UnixTimestamp,
+}
+
+/// An opportunistic cache for quotas.
+#[derive(Debug)]
+pub struct OpportunisticQuotaCache<T>
+where
+    T: std::hash::Hash + Eq,
+{
+    /// A cache for all currently consumed quotas.
+    ///
+    /// The cache is keyed with the individual quota, the value is the last known, currently
+    /// consumed amount of the quota.
+    cache: papaya::HashMap<T, CachedQuota>,
+
+    /// The amount the cache is allowed to opportunistically over-accept based on the remaining
+    /// quota.
+    ///
+    /// For example: Setting this to `10` means, if there is 100 quota remaining, the cache will
+    /// opportunistically accept the next 10 items, if there is a quota of 90 remaining, the cache
+    /// will accept the next 9 items.
+    max_over_spend_divisor: NonZeroUsize,
+
+    /// Minimum interval between vacuum of the cache.
+    vacuum_interval: Duration,
+    /// Last time then vacuum was ran.
```

Comment:
Reading the code of `try_vacuum` this seems wrong. I believe it should be
```suggestion
    /// Unix timestamp of the next time the vacuum should be run.
```